ObjectiveGit/GTRepository+Pull.m

```diff
@@ -0,0 +1,118 @@
+//
+//  GTRepository+Pull.m
+//  ObjectiveGitFramework
+//
+//  Created by Ben Chatelain on 6/17/15.
+//  Copyright © 2015 GitHub, Inc. All rights reserved.
+//
+
+#import "GTRepository+Pull.h"
+
+#import "GTCommit.h"
+#import "GTRemote.h"
+#import "GTRepository+Committing.h"
+#import "GTRepository+RemoteOperations.h"
+#import "git2/merge.h"
+
+@implementation GTRepository (Pull)
+
+#pragma mark - Pull
+
+- (BOOL)pullBranch:(GTBranch *)branch fromRemote:(GTRemote *)remote withOptions:(NSDictionary *)options
+			 error:(NSError **)error progress:(GTRemoteFetchTransferProgressBlock)progressBlock
+{
+	NSParameterAssert(branch);
+	NSParameterAssert(remote);
+
+	GTRepository *repo = remote.repository;
+
+	GTBranch *remoteBranch;
+	if (branch.branchType == GTBranchTypeLocal) {
+		BOOL success;
+		remoteBranch = [branch trackingBranchWithError:error success:&success];
+		if (!success) {
+			return NO;
+		}
+	}
+	else {
+		remoteBranch = branch;
+	}
+
+	if (![self fetchRemote:remote withOptions:options error:error progress:progressBlock]) {
+		return NO;
+	}
+
+	// Check if merge is necessary
+	GTBranch *localBranch = [repo currentBranchWithError:error];
+	if (*error) {
+		return NO;
+	}
+
+	GTCommit *localCommit = [localBranch targetCommitAndReturnError:error];
+	if (*error) {
+		return NO;
+	}
+
+	GTCommit *remoteCommit = [remoteBranch targetCommitAndReturnError:error];
+	if (*error) {
+		return NO;
+	}
+
+	if ([localCommit.SHA isEqualToString:remoteCommit.SHA]) {
+		return YES;
+	}
+
+	GTMergeAnalysis analysis = [self analyseMerge:branch fromBranch:remoteBranch error:error];
+
+	if (*error) {
+		return NO;
+	}
+
+	if (analysis & GTMergeAnalysisUpToDate) {
+		// Nothing to do
+		return YES;
+	} else if (analysis & GTMergeAnalysisFastForward) {
+		// Do FastForward
+		NSLog(@"Fast-Forward merge is not yet supported");
+		return NO;
+	} else if (analysis & GTMergeAnalysisNormal) {
+		// Do normal merge
+		GTTree *remoteTree = remoteCommit.tree;
+		NSString *message = [NSString stringWithFormat:@"Merge branch '%@'", localBranch.shortName];
+		NSArray *parents = @[ localCommit, remoteCommit ];
+		GTCommit *mergeCommit = [repo createCommitWithTree:remoteTree message:message
+												   parents:parents updatingReferenceNamed:localBranch.name
+													 error:error];
+		if (!mergeCommit) {
+			return NO;
+		}
+
+		return YES;
+	} else if (analysis & GTMergeAnalysisUnborn) {
+		// Do unborn merge
+		NSLog(@"Unborn merge is not yet supported");
+		return NO;
+	}
+
+	return NO;
+}
+
+- (GTMergeAnalysis)analyseMerge:(GTBranch *)toBranch fromBranch:(GTBranch *)fromBranch error:(NSError **)error
+{
+	git_merge_analysis_t analysis;
```

Comment:
FYI (because you had me wondering) : https://github.com/libgit2/libgit2/blob/master/src/merge.c#L2672. Not that it really matters though, it's just a default value.